x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriod.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.scalar.convert;
+
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.xpack.esql.capabilities.Validatable;
+import org.elasticsearch.xpack.esql.common.Failures;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.esql.core.type.DataType.DATE_PERIOD;
+import static org.elasticsearch.xpack.esql.core.type.DataType.isString;
+import static org.elasticsearch.xpack.esql.expression.Validations.isFoldable;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter.foldToTemporalAmount;
+
+public class ToDatePeriod extends AbstractConvertFunction implements Validatable {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        Expression.class,
+        "ToDatePeriod",
+        ToDatePeriod::new
+    );
+
+    @FunctionInfo(
+        returnType = "date_period",
+        description = "Converts an input value into a `date_period` value.",
+        examples = @Example(file = "convert", tag = "castToDatePeriod")
+    )
+    public ToDatePeriod(
+        Source source,
+        @Param(
+            name = "field",
+            type = { "date_period", "keyword", "text" },
+            description = "Input value. The input is a valid constant date period expression."
+        ) Expression v
+    ) {
+        super(source, v);
+    }
+
+    private ToDatePeriod(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected final TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        return isType(field(), dt -> isString(dt) || dt == DATE_PERIOD, sourceText(), null, "date_period or string");
+    }
+
+    @Override
+    protected Map<DataType, BuildFactory> factories() {
+        return Map.of();
+    }
```

Comment:
Thanks for catching this issue! I agree we should show the correct error messages. I looked into it, and it looks to me that this could be related to the order of validations.

I have an idea to make it show the correct message, but it modifies `resolveType()` in an unusual way. `resolveType()` usually checks if its children is resolved at the very beginning, if there is unresolved child, the rest of the validations/checks will be skipped, and it kinds of hide the error messages that we wanted. 

For this particular case, these two conversion functions don't support fields, if we add a check in `resolveType()` - if it sees a field, then return a type resolution error, the expected messages can be seen. However the check to fields need to happen before the check to its children, in order to have the expected messages returned by `Analyzer`'s `Verifier`.

```
+ curl -u elastic:password -H 'Content-Type: application/json' '127.0.0.1:9200/_query?format=txt' -d '{
    "query": "from test* | eval y = now() + to_dateperiod(x)"}
'
{"error":{"root_cause":[{"type":"verification_exception","reason":"Found 2 problems\nline 1:45: Cannot use field [x] due to ambiguities being mapped as [2] incompatible types: [keyword] in [test], [text] in [test2]\nline 1:31: argument of [to_dateperiod(x)] must be a constant, received [x]"}],"type":"verification_exception","reason":"Found 2 problems\nline 1:45: Cannot use field [x] due to ambiguities being mapped as [2] incompatible types: [keyword] in [test], [text] in [test2]\nline 1:31: argument of [to_dateperiod(x)] must be a constant, received [x]"},"status":400}%       
```

With this change, the expected message can be seen in the `VerificationException`. This is what I can think so far to make the correct message show, please let me know if there is a better way, I think we should return the expected error messages.